File: reference/provenance-ts/test/receipt.test.ts
import { test } from 'node:test'
import assert from 'node:assert/strict'
import { webcrypto } from 'node:crypto'
import { readFileSync } from 'node:fs'
import { fileURLToPath } from 'node:url'
import { dirname, join } from 'node:path'

import { emit, verify, type ReceiptPayload } from '../src/receipt.js'
import { buildChain, ChainError } from '../src/graph.js'
import { canonicalString } from '../src/canonical.js'

const subtle = webcrypto.subtle
const here = dirname(fileURLToPath(import.meta.url))

async function genKey(): Promise<webcrypto.CryptoKeyPair> {
  return (await subtle.generateKey({ name: 'Ed25519' }, true, [
    'sign',
    'verify',
  ])) as unknown as webcrypto.CryptoKeyPair
}

function basePayload(over: Partial<ReceiptPayload> = {}): ReceiptPayload {
  return {
    iat: 1700000001,
    agent_id: 'agent:test.scanner',
    agent_key_id: 'k_test01',
    operation: 'user_input',
    parents: [],
    input_hash:
      'sha256:f8c3bf62a9aa3e6fc1619c250e48abe7519373d3edf41be62eb5dc45199af2ef',
    taint: ['untrusted_user'],
    ...over,
  }
}

// ── canonical ────────────────────────────────────────────────────

test('canonical sorts keys and rejects floats', () => {
  assert.equal(canonicalString({ b: 1, a: 2 }), '{"a":2,"b":1}')
  assert.throws(() => canonicalString(1.5), /integer/)
})

// ── emit/verify roundtrip ────────────────────────────────────────

test('emit/verify roundtrip preserves payload + stable id', async () => {
  const { privateKey, publicKey } = await genKey()
  const r = await emit(basePayload(), privateKey)
  const parsed = await verify(r.jws, publicKey)
  assert.equal(parsed.payload.agent_id, 'agent:test.scanner')
  assert.equal(parsed.id, r.id)
  assert.ok(r.id.startsWith('sha256:'))
  assert.equal(parsed.payload.iss, 'raucle-detect/provenance')
  assert.equal(parsed.payload.typ, 'provenance-receipt/v1')
})

test('verify rejects a different key', async () => {
  const a = await genKey()
  const b = await genKey()
  const r = await emit(basePayload(), a.privateKey)
  await assert.rejects(() => verify(r.jws, b.publicKey), /signature invalid/)
})

test('verify rejects wrong alg', async () => {
  const { privateKey, publicKey } = await genKey()
  const r = await emit(basePayload(), privateKey)
  const [, payloadB, sigB] = r.jws.split('.')
  const badHeader = Buffer.from(
    JSON.stringify({
      alg: 'HS256',
      typ: 'provenance-receipt/v1',
      kid: 'k_test01',
      crit: ['raucle/v1'],
      'raucle/v1': 'provenance',
    }),
  )
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '')
  await assert.rejects(
    () => verify(`${badHeader}.${payloadB}.${sigB}`, publicKey),
    /unsupported alg/,
  )
})

test('verify requires crit raucle/v1', async () => {
  const { privateKey, publicKey } = await genKey()
  const r = await emit(basePayload(), privateKey)
  const [, payloadB, sigB] = r.jws.split('.')
  const badHeader = Buffer.from(
    JSON.stringify({
      alg: 'EdDSA',
      typ: 'provenance-receipt/v1',
      kid: 'k_test01',
      crit: [],
      'raucle/v1': 'provenance',
    }),
  )
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '')
  await assert.rejects(
    () => verify(`${badHeader}.${payloadB}.${sigB}`, publicKey),
    /crit/,
  )
})

// ── payload validation ───────────────────────────────────────────

test('non-user_input requires parents', async () => {
  const { privateKey } = await genKey()
  const p = basePayload({ operation: 'model_call', parents: [], model: 'test-model-v1' })
  await assert.rejects(() => emit(p, privateKey), /parent/)
})

test('rejects reserved unknown field on emit', async () => {
  const { privateKey } = await genKey()
  const p = basePayload()
  ;(p as Record<string, unknown>)['rogue'] = true
  await assert.rejects(() => emit(p, privateKey), /reserved|unknown/)
})

test('allows x_ extension field', async () => {
  const { privateKey, publicKey } = await genKey()
  const p = basePayload()
  ;(p as Record<string, unknown>)['x_trace'] = 'abc'
  const r = await emit(p, privateKey)
  const parsed = await verify(r.jws, publicKey)
  assert.equal(parsed.payload['x_trace'], 'abc')
})

// ── chain DAG + taint ────────────────────────────────────────────

test('chain validates topo + closure', async () => {
  const { privateKey } = await genKey()
  const r1 = await emit(basePayload(), privateKey)
  const r2 = await emit(
    basePayload({
      operation: 'model_call',
      parents: [r1.id],
      taint: ['untrusted_user'],
      model: 'test-model-v1',
    }),
    privateKey,
  )
  const chain = buildChain([r1, r2])
  assert.deepEqual(chain.receipts.map((r) => r.id), [r1.id, r2.id])
})

test('chain rejects topo break', async () => {
  const { privateKey } = await genKey()
  const r1 = await emit(basePayload(), privateKey)
  const r2 = await emit(
    basePayload({
      operation: 'model_call',
      parents: [r1.id],
      taint: ['untrusted_user'],
      model: 'test-model-v1',
    }),
    privateKey,
  )
  assert.throws(() => buildChain([r2, r1]), ChainError)
})

test('chain rejects silent taint loss', async () => {
  const { privateKey } = await genKey()
  const r1 = await emit(basePayload(), privateKey)
  const r2 = await emit(
    basePayload({
      operation: 'model_call',
      parents: [r1.id],
      taint: [],
      model: 'test-model-v1',
    }),
    privateKey,
  )
  assert.throws(() => buildChain([r1, r2]), /monotonicity/)
})

test('sanitisation removes tag via corpus', async () => {
  const { privateKey } = await genKey()
  const r1 = await emit(basePayload(), privateKey)
  const ok = await emit(
    basePayload({
      operation: 'sanitisation',
      parents: [r1.id],
      taint: [],
      tool: 'redactor:pii-v1',
      corpus: 'removed:untrusted_user',
    }),
    privateKey,
  )
  const chain = buildChain([r1, ok])
  assert.equal(chain.receipts.length, 2)
})

test('sanitisation undeclared drop fails', async () => {
  const { privateKey } = await genKey()
  const r1 = await emit(basePayload(), privateKey)
  const bad = await emit(
    basePayload({
      operation: 'sanitisation',
      parents: [r1.id],
      taint: [],
      tool: 'redactor:pii-v1',
      corpus: 'removed:something_else',
    }),
    privateKey,
  )
  assert.throws(() => buildChain([r1, bad]), /corpus removed-set/)
})

// ── shared cross-language conformance: the published test vectors ──

function b64uToBytes(s: string): Uint8Array {
  const pad = s.length % 4 === 0 ? '' : '='.repeat(4 - (s.length % 4))
  return new Uint8Array(Buffer.from(s.replace(/-/g, '+').replace(/_/g, '/') + pad, 'base64'))
}

async function importSeedKey(seedHex: string): Promise<webcrypto.CryptoKey> {
  const seed = Buffer.from(seedHex, 'hex')
  // PKCS#8 wrapper for a raw Ed25519 seed (RFC 8410).
  const prefix = Buffer.from('302e020100300506032b657004220420', 'hex')
  const pkcs8 = Buffer.concat([prefix, seed])
  return subtle.importKey('pkcs8', pkcs8, { name: 'Ed25519' }, false, ['sign'])
}

async function importPubPem(pem: string): Promise<webcrypto.CryptoKey> {
  const b64 = pem
    .replace(/-----BEGIN PUBLIC KEY-----/, '')
    .replace(/-----END PUBLIC KEY-----/, '')
    .replace(/\s+/g, '')
  const der = new Uint8Array(Buffer.from(b64, 'base64'))
  return subtle.importKey('spki', der, { name: 'Ed25519' }, false, ['verify'])
}

test('spec vectors: verify + byte-identical re-emit', async () => {
  const vectorsPath = join(here, '..', '..', '..', 'docs', 'spec', 'provenance', 'v1', 'test-vectors.json')
  const vf = JSON.parse(readFileSync(vectorsPath, 'utf8'))
  const priv = await importSeedKey(vf.fixed_seed_hex)
  const pub = await importPubPem(vf.public_key_pem)

  assert.ok(vf.vectors.length > 0)
  for (const v of vf.vectors) {
    // (a) Verify the published JWS + recompute its content id.
    const r = await verify(v.expected_jws, pub)
    assert.equal(r.id, v.expected_receipt_hash, `${v.name}: receipt_hash mismatch`)

    // (b) Re-emit from the fixed seed; the TS JWS + id MUST be
    //     byte-identical to the published vector.
    const payloadObj = JSON.parse(
      new TextDecoder().decode(b64uToBytes(v.expected_jws.split('.')[1])),
    ) as ReceiptPayload
    const emitted = await emit(payloadObj, priv)
    assert.equal(emitted.jws, v.expected_jws, `${v.name}: emitted JWS differs`)
    assert.equal(emitted.id, v.expected_receipt_hash, `${v.name}: emitted id differs`)
  }
})

File: reference/provenance-go/receipt_test.go
package provenance

import (
	"crypto/ed25519"
	"crypto/x509"
	"encoding/base64"
	"encoding/json"
	"encoding/pem"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func parseEd25519PubPEM(p string) (ed25519.PublicKey, error) {
	block, _ := pem.Decode([]byte(p))
	if block == nil {
		return nil, fmt.Errorf("no PEM block")
	}
	key, err := x509.ParsePKIXPublicKey(block.Bytes)
	if err != nil {
		return nil, err
	}
	pk, ok := key.(ed25519.PublicKey)
	if !ok {
		return nil, fmt.Errorf("not an ed25519 key")
	}
	return pk, nil
}

func basePayload() Payload {
	return Payload{
		Iat:        1700000001,
		AgentID:    "agent:test.scanner",
		AgentKeyID: "k_test01",
		Operation:  "user_input",
		Parents:    nil,
		InputHash:  "sha256:f8c3bf62a9aa3e6fc1619c250e48abe7519373d3edf41be62eb5dc45199af2ef",
		Taint:      []string{"untrusted_user"},
	}
}

func mustKeys(t *testing.T) (ed25519.PublicKey, ed25519.PrivateKey) {
	pub, priv, err := ed25519.GenerateKey(nil)
	if err != nil {
		t.Fatal(err)
	}
	return pub, priv
}

func TestEmitVerifyRoundtrip(t *testing.T) {
	pub, priv := mustKeys(t)
	r, err := Emit(basePayload(), priv)
	if err != nil {
		t.Fatal(err)
	}
	parsed, err := Verify(r.JWS, pub)
	if err != nil {
		t.Fatal(err)
	}
	if parsed.Payload.AgentID != "agent:test.scanner" {
		t.Fatalf("agent_id mismatch: %s", parsed.Payload.AgentID)
	}
	if parsed.ID != r.ID || !strings.HasPrefix(r.ID, "sha256:") {
		t.Fatalf("id mismatch/prefix: %s", parsed.ID)
	}
}

func TestVerifyRejectsDifferentKey(t *testing.T) {
	_, priv := mustKeys(t)
	otherPub, _ := mustKeys(t)
	r, _ := Emit(basePayload(), priv)
	if _, err := Verify(r.JWS, otherPub); err == nil {
		t.Fatal("expected signature failure")
	}
}

func TestVerifyRejectsWrongAlg(t *testing.T) {
	pub, priv := mustKeys(t)
	r, _ := Emit(basePayload(), priv)
	parts := strings.Split(r.JWS, ".")
	badHeader, _ := json.Marshal(map[string]any{
		"alg": "HS256", "typ": jwsTyp, "kid": "k_test01",
		"crit": []string{"raucle/v1"}, "raucle/v1": "provenance",
	})
	hb := base64.RawURLEncoding.EncodeToString(badHeader)
	if _, err := Verify(hb+"."+parts[1]+"."+parts[2], pub); err == nil {
		t.Fatal("expected alg rejection")
	}
}

func TestVerifyRequiresCrit(t *testing.T) {
	pub, priv := mustKeys(t)
	r, _ := Emit(basePayload(), priv)
	parts := strings.Split(r.JWS, ".")
	badHeader, _ := json.Marshal(map[string]any{
		"alg": "EdDSA", "typ": jwsTyp, "kid": "k_test01",
		"crit": []string{}, "raucle/v1": "provenance",
	})
	hb := base64.RawURLEncoding.EncodeToString(badHeader)
	if _, err := Verify(hb+"."+parts[1]+"."+parts[2], pub); err == nil {
		t.Fatal("expected crit rejection")
	}
}

func TestNonUserInputRequiresParents(t *testing.T) {
	_, priv := mustKeys(t)
	p := basePayload()
	p.Operation = "model_call"
	p.Model = "test-model-v1"
	if _, err := Emit(p, priv); err == nil {
		t.Fatal("expected parent requirement")
	}
}

func TestRejectsUnknownPayloadField(t *testing.T) {
	if _, err := payloadFromMap(map[string]any{
		"typ": typ, "operation": "user_input", "rogue": true,
	}); err == nil {
		t.Fatal("expected unknown-field rejection")
	}
}

func TestRejectsMissingTyp(t *testing.T) {
	if _, err := payloadFromMap(map[string]any{
		"operation": "user_input",
	}); err == nil {
		t.Fatal("expected typ rejection")
	}
}

func TestChainTopoAndClosure(t *testing.T) {
	_, priv := mustKeys(t)
	r1, _ := Emit(basePayload(), priv)
	p2 := basePayload()
	p2.Operation = "model_call"
	p2.Parents = []string{r1.ID}
	p2.Model = "test-model-v1"
	r2, _ := Emit(p2, priv)
	c, err := BuildChain([]Receipt{r1, r2})
	if err != nil {
		t.Fatal(err)
	}
	if len(c.Receipts) != 2 {
		t.Fatalf("expected 2 receipts, got %d", len(c.Receipts))
	}
}

func TestChainRejectsTopoBreak(t *testing.T) {
	_, priv := mustKeys(t)
	r1, _ := Emit(basePayload(), priv)
	p2 := basePayload()
	p2.Operation = "model_call"
	p2.Parents = []string{r1.ID}
	p2.Model = "test-model-v1"
	r2, _ := Emit(p2, priv)
	if _, err := BuildChain([]Receipt{r2, r1}); err == nil {
		t.Fatal("expected topo violation")
	}
}

func TestChainRejectsSilentTaintLoss(t *testing.T) {
	_, priv := mustKeys(t)
	r1, _ := Emit(basePayload(), priv)
	p2 := basePayload()
	p2.Operation = "model_call"
	p2.Parents = []string{r1.ID}
	p2.Taint = nil
	p2.Model = "test-model-v1"
	r2, _ := Emit(p2, priv)
	if _, err := BuildChain([]Receipt{r1, r2}); err == nil {
		t.Fatal("expected taint monotonicity violation")
	}
}

func TestSanitisationRemovesTagViaCorpus(t *testing.T) {
	_, priv := mustKeys(t)
	r1, _ := Emit(basePayload(), priv)
	p2 := basePayload()
	p2.Operation = "sanitisation"
	p2.Parents = []string{r1.ID}
	p2.Taint = nil
	p2.Tool = "redactor:pii-v1"
	p2.Corpus = "removed:untrusted_user"
	p2.OutputHash = "sha256:e54b74eb9192b48055c48d2062bffdd23469ef7d70f960ff1293a47f86c8eba2"
	r2, _ := Emit(p2, priv)
	if _, err := BuildChain([]Receipt{r1, r2}); err != nil {
		t.Fatal(err)
	}
}

func TestSanitisationUndeclaredDropFails(t *testing.T) {
	_, priv := mustKeys(t)
	r1, _ := Emit(basePayload(), priv)
	p2 := basePayload()
	p2.Operation = "sanitisation"
	p2.Parents = []string{r1.ID}
	p2.Taint = nil
	p2.Tool = "redactor:pii-v1"
	p2.Corpus = "removed:something_else"
	p2.OutputHash = "sha256:e54b74eb9192b48055c48d2062bffdd23469ef7d70f960ff1293a47f86c8eba2"
	r2, _ := Emit(p2, priv)
	if _, err := BuildChain([]Receipt{r1, r2}); err == nil {
		t.Fatal("expected sanitisation taint mismatch")
	}
}

// TestCanonicalParity locks the canonical-JSON output to the exact
// bytes the Python + TS reference encoders produce for the same object.
func TestCanonicalParity(t *testing.T) {
	got, err := canonicalEncode(map[string]any{
		"iss": "x", "iat": 1, "parents": []any{"a", "b"},
		"taint": []any{"a_t", "z_t"},
	})
	if err != nil {
		t.Fatal(err)
	}
	want := `{"iat":1,"iss":"x","parents":["a","b"],"taint":["a_t","z_t"]}`
	if string(got) != want {
		t.Fatalf("canonical mismatch:\n got=%s\nwant=%s", got, want)
	}
}

// ── shared cross-language conformance: the published test vectors ──

type vectorFile struct {
	PublicKeyPEM string `json:"public_key_pem"`
	FixedSeedHex string `json:"fixed_seed_hex"`
	Vectors      []struct {
		Name            string `json:"name"`
		ExpectedJWS     string `json:"expected_jws"`
		ExpectedReceipt string `json:"expected_receipt_hash"`
	} `json:"vectors"`
}

func loadVectors(t *testing.T) (vectorFile, ed25519.PublicKey) {
	t.Helper()
	path := filepath.Join("..", "..", "docs", "spec", "provenance", "v1", "test-vectors.json")
	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("read vectors: %v", err)
	}
	var vf vectorFile
	if err := json.Unmarshal(raw, &vf); err != nil {
		t.Fatal(err)
	}
	pub, err := parseEd25519PubPEM(vf.PublicKeyPEM)
	if err != nil {
		t.Fatalf("parse pubkey: %v", err)
	}
	return vf, pub
}

// payloadFromJWS decodes (without verifying) the payload object from a
// compact JWS into a Payload, so we can re-emit it from the fixed seed.
func payloadFromJWS(t *testing.T, jws string) Payload {
	t.Helper()
	parts := strings.Split(jws, ".")
	pb, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		t.Fatal(err)
	}
	var m map[string]any
	if err := json.Unmarshal(pb, &m); err != nil {
		t.Fatal(err)
	}
	p, err := payloadFromMap(m)
	if err != nil {
		t.Fatal(err)
	}
	return p
}

func TestSpecVectors(t *testing.T) {
	vf, pub := loadVectors(t)
	if len(vf.Vectors) == 0 {
		t.Fatal("no vectors loaded")
	}
	seed, err := hexDecode(vf.FixedSeedHex)
	if err != nil {
		t.Fatal(err)
	}
	priv := ed25519.NewKeyFromSeed(seed)

	for _, v := range vf.Vectors {
		t.Run(v.Name, func(t *testing.T) {
			// (a) Verify the published JWS + recompute its content id.
			r, err := Verify(v.ExpectedJWS, pub)
			if err != nil {
				t.Fatalf("verify failed: %v", err)
			}
			if r.ID != v.ExpectedReceipt {
				t.Fatalf("receipt_hash mismatch:\n got=%s\nwant=%s", r.ID, v.ExpectedReceipt)
			}
			// (b) Re-emit from the fixed seed: the Go-produced JWS and id
			//     MUST be byte-identical to the published vector.
			p := payloadFromJWS(t, v.ExpectedJWS)
			emitted, err := Emit(p, priv)
			if err != nil {
				t.Fatalf("emit failed: %v", err)
			}
			if emitted.JWS != v.ExpectedJWS {
				t.Fatalf("emitted JWS differs:\n got=%s\nwant=%s", emitted.JWS, v.ExpectedJWS)
			}
			if emitted.ID != v.ExpectedReceipt {
				t.Fatalf("emitted id differs:\n got=%s\nwant=%s", emitted.ID, v.ExpectedReceipt)
			}
		})
	}
}

func hexDecode(s string) ([]byte, error) {
	out := make([]byte, len(s)/2)
	for i := 0; i < len(out); i++ {
		var b int
		_, err := fmt.Sscanf(s[2*i:2*i+2], "%02x", &b)
		if err != nil {
			return nil, err
		}
		out[i] = byte(b)
	}
	return out, nil
}
